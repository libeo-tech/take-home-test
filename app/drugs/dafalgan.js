import {DrugModel} from "./models/drug";

export class Dafalgan extends DrugModel {
  constructor(expiresIn, benefit) {
    super(DrugModel.drugs.DAFALGAN, expiresIn, benefit)
  }

  updateBenefitValue(data) {
    try {
      DrugModel.validate(data);

      if (data.benefit == DrugModel.minBenefitValue) return data;

      let benefitReduction = 0;
      if (data.benefit > DrugModel.minBenefitValue) benefitReduction += 2;
      if (data.expiresIn <= 0) benefitReduction += 2;

      data.benefit = data.benefit - benefitReduction;

      if (data.benefit < DrugModel.minBenefitValue) data.benefit = DrugModel.minBenefitValue;

      return data;
    } catch (error) {
      throw error;
    }
  }

  updateExpiresInValue(data) {
    try {
      DrugModel.validate(data);

      return DrugModel.updateExpiresInValue(data);
    } catch (error) {
      throw error;
    }
  }
}
