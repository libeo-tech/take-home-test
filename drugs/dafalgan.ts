import { Drug, DrugName } from "./drug";

export class Dafalgan extends Drug {
    constructor(expiresIn: number, benefit: number) {
        super(DrugName.DAFALGAN, expiresIn, benefit);
    }
    
    public updateDrug(): void {
        if (this.isExpired) {
            this.decreaseBenefit(2);
            return;
        }
        this.decreaseExpiresIn();
        this.decreaseBenefit(2);
    }
}