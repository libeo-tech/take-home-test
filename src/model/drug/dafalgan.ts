import { DrugName } from '../../type/DrugName';
import { Drug } from './drug';

export class Dafalgan extends Drug {
  constructor(expiresIn: number, benefit: number) {
    super(DrugName.DAFALGAN, expiresIn, benefit);
  }

  public updateValues(): this {
    this.benefit -= this.isExpired() ? 4 : 2;
    this.expiresIn -= 1;

    return this;
  }
}
