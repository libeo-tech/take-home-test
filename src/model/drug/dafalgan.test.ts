import { Dafalgan } from './dafalgan';

describe('Dafalgan', () => {
  let dafalgan: Dafalgan;

  beforeEach(() => {
    dafalgan = new Dafalgan(30, 5);
  });

  it('should not allow for a negative benefit', () => {
    dafalgan.benefit = -5;
    expect(dafalgan).toMatchSnapshot();
  });

  it('should not allow for a benefit greater than 50', () => {
    dafalgan.benefit = 51;
    expect(dafalgan).toMatchSnapshot();
  });

  it('should return expiration status', () => {
    expect(dafalgan.isExpired()).toBe(false);
    dafalgan.expiresIn = 0;
    expect(dafalgan.isExpired()).toBe(true);
  });

  it('should decrease expiresIn and benefit over time', () => {
    dafalgan.updateValues();
    expect(dafalgan).toMatchSnapshot();
  });

  it('should decrease benefit twice as much if expired', () => {
    dafalgan.expiresIn = 0;
    dafalgan.updateValues();
    expect(dafalgan).toMatchSnapshot();
  });

  it('should be serializable to JSON', () => {
    expect(dafalgan.toJSON()).toMatchSnapshot();
  });

  it('should be serializable to string', () => {
    expect(dafalgan.toString()).toMatchSnapshot();
  });
});
