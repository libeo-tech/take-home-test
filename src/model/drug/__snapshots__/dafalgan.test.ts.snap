// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dafalgan should be serializable to JSON 1`] = `
Object {
  "benefit": 5,
  "expiresIn": 30,
  "name": "Dafalgan",
}
`;

exports[`Dafalgan should be serializable to string 1`] = `"{\\"name\\":\\"Dafalgan\\",\\"expiresIn\\":30,\\"benefit\\":5}"`;

exports[`Dafalgan should decrease benefit twice as much if expired 1`] = `
Object {
  "benefit": 1,
  "expiresIn": -1,
  "name": "Dafalgan",
}
`;

exports[`Dafalgan should decrease expiresIn and benefit over time 1`] = `
Object {
  "benefit": 3,
  "expiresIn": 29,
  "name": "Dafalgan",
}
`;

exports[`Dafalgan should not allow for a benefit greater than 50 1`] = `
Object {
  "benefit": 50,
  "expiresIn": 30,
  "name": "Dafalgan",
}
`;

exports[`Dafalgan should not allow for a negative benefit 1`] = `
Object {
  "benefit": 0,
  "expiresIn": 30,
  "name": "Dafalgan",
}
`;
